cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -382,119 +412,60 @@ impl YaccParser {
-        if let Some(j) = self.lookahead_is("%grmtools", i) {
-            i = self.parse_ws(j, true)?;
-            if let Some(j) = self.lookahead_is("{", i) {
-                i = self.parse_ws(j, true)?;
-                while self.lookahead_is("}", i).is_none() {
-                    let key_start_pos = i;
-                    let (key_end_pos, key) = self.parse_name(i)?;
-                    i = self.parse_ws(key_end_pos, false)?;
-                    if key == "yacckind" {
-                        if let Some(j) = self.lookahead_is(":", i) {
-                            i = self.parse_ws(j, false)?;
-                        } else {
-                            return Err(YaccGrammarError {
-                                kind: YaccGrammarErrorKind::InvalidGrmtoolsHeaderKey,
-                                spans: vec![Span::new(i, i)],
-                            });
-                        }
-                        let val_end_pos = self.parse_yacckind(i, update_yacc_kind)?;
-                        if let Some(orig) = yacc_kind_key_span {
-                            let dupe = Span::new(key_start_pos, key_end_pos);
-                            return Err(YaccGrammarError {
-                                kind: YaccGrammarErrorKind::DuplicateGrmtoolsHeaderKey,
-                                spans: vec![orig, dupe],
-                            });
-                        } else {
-                            yacc_kind_key_span = Some(Span::new(key_start_pos, key_end_pos));
-                            i = self.parse_ws(val_end_pos, true)?;
-                        }
-                        if let Some(j) = self.lookahead_is(",", i) {
-                            i = self.parse_ws(j, true)?;
-                            continue;
-                        } else {
-                            break;
+                Value::Setting(Setting::Num(_, span)) => {
+                    return Err(YaccGrammarError {
+                        kind: YaccGrammarErrorKind::InvalidYaccKind,
+                        spans: vec![*span],
+                    })
+                }
+                Value::Setting(Setting::Unitary((namespace, (value, _)))) => {
+                    for (yk_str, yk) in [
+                        ("grmtools".to_string(), YaccKind::Grmtools),
+                        ("eco".to_string(), YaccKind::Eco),
+                    ] {
+                        if (namespace.as_ref().map(|(x, _)| x) == Some(&yacckind_ns)
+                            || namespace.is_none())
+                            && *value == yk_str
+                        {
+                            self.yacc_kind = Some(yk);
                         }
-                    } else {
-                        return Err(YaccGrammarError {
-                            kind: YaccGrammarErrorKind::InvalidGrmtoolsHeaderKey,
-                            spans: vec![Span::new(key_start_pos, key_end_pos)],
-                        });
                     }
```

Comment:
I hadn't settled here on how best to return an error on invalid `yacckind` value in this case or the `Constructor` case. At the very least this loop should stop after the first match... 

I guess i'll try `let x = for ... {  if ... {break Some(yk) } else { None } };` then throw an error on None.

Reply:
I did a pass over this in [23a1391](https://github.com/softdevteam/grmtools/pull/546/commits/23a1391cd40eed077eca29612d824ed17d11fec7)

Having rewritten it, I feel like the error handling here is probably more robust, in the sense
that previously we were using `starts_with` in many cases, which the `invalid_yacckind` tests I added should cover.

That said I wouldn't be suprised if this could be drastically improved somehow by avoiding `loop`.